### ISSUE TITLE
frozen and applySnapshot [bug]

### ISSUE THREAD

**Issue by maxguzenski** [mobxjs/mobx-state-tree#634]
When use  applySnapshot it set a optional t.frozen to undefined...
Is this expected ?

```javascript
const Model = types.model('model', {
  field: t.optional(t.frozen, {})
})

const m = Model.create({})
m.field == {} 

applySnapshot(m, {})
m.field == undefined
```

**Comment by elite174** [comment]
Try this.
https://jsfiddle.net/43a9bepv/

**Comment by maxguzenski** [comment]
Your link is using a old mobx-state-tree version, where there is no bug... try this:

https://mattiamanzati.github.io/mobx-state-tree-playground/#src=import%20React%20from%20%22react%22%0Aimport%20%7B%20types%2C%20applySnapshot%20%7D%20from%20%22mobx-state-tree%22%0A%0Aconst%20Model%20%3D%20types.model('model'%2C%20%7B%0A%20%20field%3A%20types.optional(types.frozen%2C%20%7B%7D)%0A%7D)%0A%0Aconst%20m%20%3D%20Model.create(%7B%7D)%0Aconsole.log(m.field%20%3D%3D%20%7B%7D)%20%0A%0AapplySnapshot(m%2C%20%7B%7D)%0Aconsole.log(m.field%20%3D%3D%20undefined)%0A%2F*%20const%20m%20%3D%20Model.create(%7B%7D)%0Aconsole.log(m.field%20%3D%3D%20%7B%0AapplySnapshot(m%2C%20%7B%7D)%0Aconsole.log(getSnapshot(m))%0Aconsole.log(m.field%20%3D%3D%20undefined)%20*%2F

**Comment by aretecode** [comment]
this is the same issue I saw here https://github.com/mobxjs/mobx-state-tree/issues/347

**Comment by robinfehr** [comment]
PR welcome. it's now on the todo list.

**Comment by SirbyAlive** [comment]
Hey guys,

I've stumbled upon this bug.
The issue does not seem to be with the ```types.frozen``` but with the ```types.optional```. Since ```frozen``` can be anything, its typecheck is always true even for ```undefined``` and it does not fallback on the default value during reconciliation.

Quick fix I used is to declare a custom optional model like:
```
function optional<S, T>(type: IType <S, T>, defaultValue: S): IType<S, T> {
  return types.custom({
    getValidationMessage: () => null,
    isTargetType: (self: any) => self !== undefined,
    name: 'optional',
    fromSnapshot(snap: any) {
        return snap !== undefined
          ? type.create(snap)
          : type.create(defaultValue);
    },
    toSnapshot: (self: any) => self,
  });
}
```

**Comment by mweststrate** [comment]
Fix released as 2.1.0
